### PR TITLE
Time - Quin 

### PR DESCRIPTION
**Pull request by quinqu**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method to validate if a user had already voted on a certain piece of media 
Describe how you approached testing that model method. What edge cases did you come up with? |  I just check if my "vote container" had the user ID of the user that was trying to upvote it. | 
What are `session` and `flash`? What is the difference between them? | session stores the users information and flash it a way to display a message for one request cycle or the next (depending on of you use flash or flash.now)
What was one thing that you gained more clarity on through this assignment? | I understood how many to many relationships worked more, and actually all of the relationships in my app but mostly many to many. 
What is the Heroku URL of your deployed application? | https://ranker-rails.herokuapp.com/

### CODE REVIEW

**Comment by quinqu** [comment]
Hello Reviewer, 

I just wanted to clarify my database schema was not set up the way I think ya'll wanted us to set it up. I thought it we could design however we wanted to at first, however when I looked at the rubric I noticed that work should have many votes. My schema says work has one vote and actually vote doesn't have a very descriptive name, it should be 'vote tally' or something. Everything works, but just a little differently behind the scenes :)

**Comment by quinqu** [comment]
Also, user join date does not show on the heroku app, but it shows on my local host

File: test/controllers/top_media_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe TopMediaController do
```
Comment (CheezItMan) [inline test/controllers/top_media_controller_test.rb:3]: Just noting no test here.

File: test/controllers/users_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      post login_path, params: user_hash
+      post logout_path
+
+      expect(session[:user_id]).must_be_nil
+
+    end
+  end 
+
+  describe " current " do 
+    it "will direct to current user" do
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:54]: You should test this with users and without any users.

File: test/controllers/votes_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+require "test_helper"
+
+describe VotesController do
+ it "will add user to vote" do
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:4]: You should also try voting for the same work twice and voting without being logged in.

File: test/controllers/votes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  
+  user_hash = {
+    user: {
+      username: "Vote User"
+    }
+  }
+
+  post login_path, params: user_hash
+  get upvote_path(vote.id)
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:16]: Why is upvoting a `get` request?

File: test/controllers/works_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      get new_work_path 
+
+      must_respond_with :success
+
+    end 
+  end 
+
+
+  describe "create" do 
+    it "can create a new work" do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:29]: You should also have failure cases.

File: test/controllers/works_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    it "will respond with redirect if nonexistent work" do
+      login()
+      get edit_work_path(-100)
+      must_respond_with :not_found
+    end
+  end 
+
+  describe "update" do 
+    it "will update el work " do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:82]: You should also test this with a failed update.

File: test/controllers/works_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    patch work_path(work.id), params: update_work
+  
+    expect(Work.first.title).must_equal update_work[:work][:title]
+    expect(Work.first.year).must_equal update_work[:work][:year]
+    end 
+  end 
+
+
+  describe "delete" do 
+    it "will delete work" do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:104]: You should also have a test trying to delete a nonexistant work.

File: test/models/vote_test.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+require "test_helper"
+
+describe Vote do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:3]: Noting that there are no tests here.

File: test/models/work_test.rb [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  it "will have the required fields" do
+    #Arrange
+    new_work.save
+    work = Work.first
+    #Act
+    [:title, :category, :year, :description, :creator].each do |field|
+      #Assert
+      expect(work).must_respond_to field
```
Comment (CheezItMan) [inline test/models/work_test.rb:21]: I like this loop to test required fields.

File: test/models/work_test.rb [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    it "will not update if there is a blank input" do 
+      work = Work.new()
+      saved = work.save
+      expect(saved).must_equal false
+    end 
+  end 
+
+  describe "work model methods" do 
+    it "will find the top media " do
```
Comment (CheezItMan) [inline test/models/work_test.rb:45]: You should test this method when works have no votes, when they do and when there are no works.

File: config/routes.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+
+Rails.application.routes.draw do
+
+  get "top_media/index"
```
Comment (CheezItMan) [inline config/routes.rb:4]: You don't need this route if you're using this controller/method for the homepage.

File: config/routes.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+  
+  post "/login", to: "users#login"
+  
+  post "/logout", to: "users#logout", as: "logout"
+
+  get "/users/current", to: "users#current", as: "current_user"
+  
+  resources :works 
+
+  resources :users
```
Comment (CheezItMan) [inline config/routes.rb:18]: Do you need all the CRUD routes for users?

File: app/controllers/votes_controller.rb [hunk 13]
```diff
@@ -0,0 +1,19 @@
+class VotesController < ApplicationController
+
+  def upvote 
+    @vote = Vote.find_by(id: params[:format])
```
Comment (CheezItMan) [inline app/controllers/votes_controller.rb:4]: You're finding a vote here, not creating one with a work and a user.  This is a very unusual solution it works, but it means that for every work you're creating a vote with a list of users.

File: app/controllers/works_controller.rb [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  end 
+
+  
+  def create 
+    @work = Work.new(work_params)
+    if @work.save 
+      redirect_to works_path
+      return 
+    else
+      render :new
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:28]: You should include a status code here.

```suggestion
      render :new, status: :bad_request
```

File: app/controllers/works_controller.rb [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+      return 
+    end 
+  end 
+
+  def edit 
+    if @work.nil?
+      head :not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:35]: In general you should use a redirect here, because it's a bad user experience to confront them with a blank screen.  Now that you know enough, transition off `head`.

File: app/controllers/works_controller.rb [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+  def update 
+    if @work.nil? 
+      head :not_found
+      return 
+    elsif @work.update(work_params)
+      flash[:welcome] = "Successfully Updated"
+      redirect_to work_path
+      return 
+    else
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:49]:
```suggestion
      render :edit, status: :bad_request
```

File: app/models/work.rb [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  validates :year, presence: true
+  validates :description, presence: true
+ 
+  def create_vote
+    vote = Vote.create(work_id: self.id)
+  end  
+
+  def self.find_top_amount(type)
+    works = Work.where(category: type)
+    works = works.sort_by { |work| work.vote.users.length }
```
Comment (CheezItMan) [inline app/models/work.rb:16]: I just want to make the point that it's more efficient to use the `.order` method to sort the list of works and let Postgres do the sorting.  It's faster than retrieving all the works and then sorting them.

File: app/views/works/index.html.erb [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+  <h3 class="works-header"> List of All Works </h3> 
+
+  <h3>Albums </h3>
+  <%= render partial: "shared/list", locals: { type: @albums } %>
+
+  <h3>Books </h3>
+  <%= render partial: "shared/list", locals: { type: @books } %>
+
+  <h3>Movies </h3>
+  <%= render partial: "shared/list", locals: { type: @movies } %>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:11]: Love these partials.

File: app/controllers/works_controller.rb [hunk 19]
```diff
@@ -0,0 +1,76 @@
+class WorksController < ApplicationController
+  before_action :find_work, except: [:index, :new, :create]
+  before_action :require_login, only: [:new, :create, :edit, :update]
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:3]: 👍

**Comment by CheezItMan** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️, you've got the bonus authorization to create a work
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ⚠ You've got an extra route and you redefined the works routes twice.
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️, You have a unique solution here to find 
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️, not in application controller, but it works.
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️, see my comments
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️, see my comments
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️, see my comments

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅




## Summary


Overall this is well done.  You do have some further work to do in terms of testing.  Your solution in terms of database structures is unusual, but it does work.  You don't have the seeds files working, which is unfortunate.  Otherwise this is a good solid submission.
